### DISCUSSION TITLE
Is there BIRT API Doc (Javadoc)?

### DISCUSSION THREAD

**Discussion by 4r73m190r0s** [eclipse-birt/birt#1602] [Q&A]
I know this is very basic question, but just can't seem to find it anywhere. I'm building a BIRT report and can't find any info on some of the methods I want to use.

**Comment by dariobella** [comment]
I don't believe there is an up-to-date javadoc available online. But i found a javadoc of different packages from version 4.4.1:

http://otadocs.opentext.com/documentation/JAVADOCS/BIRT-4.4.1/org.eclipse.birt.doc.isv/enginescript/api/overview-summary.html
http://otadocs.opentext.com/documentation/JAVADOCS/BIRT-4.4.1/org.eclipse.birt.chart.doc.isv/chart/api/overview-summary.html
http://otadocs.opentext.com/documentation/JAVADOCS/BIRT-4.4.1/org.eclipse.birt.doc.isv/engine/api/overview-summary.html

The way I found these is basically just searching for "BIRT" + the exact name of the interface, class or method I needed, and normally there's an opentext result somewhere. Sometimes they are from an earlier version, but you can always replace the version with "4.4.1" in the URL. Obviously it's not the most recent one so some things might be different, but it's better than nothing :)

**Comment by hvbtup** [comment]
Well, you can also look directly at the Java Source code.
But, on a high level, including examples, I think the web page shows how to integrate BIRT into your own code (not with Maven, unfortunately) and how to _run_ reports. 
See https://eclipse-birt.github.io/birt-website/docs/integrating/reapi#irunandrendertask
This example generates HTML output. Generating PDF works very similar.

But you say you are _building_ a report. So maybe you are more interested in BIRT's scripting capabilities (that is, custom logic inside a report). Here, Javadoc is not appropriate, because scripting is done with Javascript.
Basically, you can explore what methods (called methods, triggers or events in other software) are available for each item by selecting the item and going to the Script page. The most important ones are:

* initialize (at report level): Here you should define Javascript functions (and maybe global variables, but also look at the dedicated "Variables" in the outline).
* onCreate (for layout items): This is basically the last chance to modify the visual appearance dynamically, in case you want to do something which cannot be done using highlight rules.
* beforeOpen event (for datasets): This is the last chance to overwrite input parameters for the query or to change the queryText itself.
* onFetch event (for datasets): For example, you you set a global variable dataFound = true here (and use it for a layout item showing "No Data Found" at the end of the report if it's not true then).
* Scripted DataSets have additional events. They are a bit more complicated.

I remember, years ago there was a web page including a picture somewhere on Actuate's web page which explained all these Javascript events in more detail, but I cannot find it now.
